// @vsintellicode/typescript-intellicode-plugin/index.js
/*! Copyright (c) Microsoft Corporation. All rights reserved. */
module.exports=function(e){var n={};function t(o){if(n[o])return n[o].exports;var i=n[o]={i:o,l:!1,exports:{}};return e[o].call(i.exports,i,i.exports,t),i.l=!0,i.exports}return t.m=e,t.c=n,t.d=function(e,n,o){t.o(e,n)||Object.defineProperty(e,n,{enumerable:!0,get:o})},t.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.t=function(e,n){if(1&n&&(e=t(e)),8&n)return e;if(4&n&&"object"==typeof e&&e&&e.__esModule)return e;var o=Object.create(null);if(t.r(o),Object.defineProperty(o,"default",{enumerable:!0,value:e}),2&n&&"string"!=typeof e)for(var i in e)t.d(o,i,function(n){return e[n]}.bind(null,i));return o},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p="",t(t.s=0)}([function(e,n,t){"use strict";var o=this&&this.__importStar||function(e){if(e&&e.__esModule)return e;var n={};if(null!=e)for(var t in e)Object.hasOwnProperty.call(e,t)&&(n[t]=e[t]);return n.default=e,n};const i=o(t(1)),r=o(t(3)),s=o(t(4)),a=o(t(5)),c=o(t(6)),u=1e6,l=1e9;let f,d,p=!1,m=500,y=void 0;function g(e,n,t){const o=S(...process.hrtime(e));n&&"None"===n.failureReason?t.metadata={type:n.foundNode.typeName,inIf:n.foundNode.isInConditional,patterns:n.sequences,computationTimeMS:o,failureReason:"None"}:t.metadata={computationTimeMS:o,failureReason:n?n.failureReason:"LookupKeyGenerationTimeout"}}function h(e,n){e.languageServiceHost.log&&e.languageServiceHost.log(`IntelliCode plugin: ${n}`)}function S(e,n){return(e*l+n)/u}class x{constructor(e,n,t){this.ts=e,this.hostCancellationToken=n,this.timeoutMS=t,this.startTime=Date.now()}throwIfCancellationRequested(){if(this.isHostTokenCanceled())throw new N(!0);if(p&&this.isLocalTokenCanceled())throw new N(!1)}isHostTokenCanceled(){return this.hostCancellationToken.isCancellationRequested()}isLocalTokenCanceled(){return Math.abs(Date.now()-this.startTime)>=this.timeoutMS}}class N{constructor(e){this.canceledByHostToken=e}}e.exports=function(e){const n=e.typescript;let t;return{create:function(e){const u=Object.create(null);for(let n of Object.keys(e.languageService)){const t=e.languageService[n];t&&(u[n]=t.bind(e.languageService))}function l(){if(!y&&f)try{const n=s.readFileSync(f),t=Buffer.from([214,105,172,23,156,35,36,45,137,37,195,176,189,110,127,78,69,171,205,160,3,202,114,220,252,109,233,221,252,181,45,47]),o=Buffer.from([60,12,128,217,163,220,176,146,213,18,190,59,212,88,192,21]),i=r.createDecipheriv("aes-256-cbc",t,o),c=Buffer.concat([i.update(n),i.final()]),u=a.gunzipSync(c);y=JSON.parse(u.toString("utf-8")),d="typescript_"+(y.Version||"unknown"),h(e,`successfully loaded model from ${f}`)}catch(n){h(e,`failed to load model from ${f}, exception: ${n}`)}return y}return o(e.config),t=e.languageServiceHost,h(e,`starting configuration: ${JSON.stringify(e.config)}`),u.getCompletionsAtPosition=(t,o,r)=>(function(e,n,t,o,r,s){const a=n.languageService.getCompletionsAtPosition(o,r,s);if(!a||!a.isMemberCompletion)return a;const u=n.project.getCancellationToken(),l=new x(e,u,m),f=process.hrtime(),y=n.languageService.getProgram(),T=y?y.getSourceFile(o):void 0;if(a.entries.length>0||!T||!c.isInString(e,T,r)){if(p){try{l.throwIfCancellationRequested();const e=v(l);g(f,e,a)}catch(n){if(!(n instanceof N))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException;g(f,void 0,a)}return a}const o=function(e){let o=void 0,i=[];const r=t(),s=r&&r.model;if(!s)return{recommendedWords:[],failureReason:"NoModelAvailable"};const a=v(e);if(!a.foundNode||!a.sequences)return{recommendedWords:[],failureReason:a.failureReason};const c=s[a.foundNode.typeName];if(c){h(n,"found some suggestion stuff");for(const t of a.sequences){e.throwIfCancellationRequested();const r=c[t],s=r&&(a.foundNode.isInConditional?r[1]:r[0]);if(s){o=t,h(n,`found suggestions for sequence ${t} : ${s.join(",")}`),i.push(...s.slice(0,5));break}}}return{typeName:a.foundNode.typeName,inConditional:a.foundNode.isInConditional,sequenceString:o,recommendedWords:i,failureReason:"None"}}(l),i=S(...process.hrtime(f));let r="None"===o.failureReason?"NotInModel":o.failureReason,s=[];o.recommendedWords.length>0&&(r="NotInIntersection",a.entries=a.entries.filter(e=>!e.replacementSpan||!e.insertText||"["!==e.insertText[0]),a.entries.some(e=>"0"===e.sortText)?o.recommendedWords.forEach((e,t)=>{const o=a.entries.findIndex(n=>n.name===e);if(o>=0){const e=a.entries[o];h(n,`marking recommendation ${e.name}`),r="None",k(e,t),s.push(e),a.entries.splice(o,1)}}):(h(n,"no strong recommendations, adding entries"),a.entries=a.entries.filter(e=>!o.recommendedWords.some(n=>e.name===n)),s=o.recommendedWords.map((n,t)=>{const o={name:n,kind:e.ScriptElementKind.memberVariableElement,sortText:"0"};return k(o,t),o})),a.entries=s.concat(a.entries));const c=function(e,n,t){return{Id:Date.now(),ModelVersion:d||"typescript_unknown",Count:0,FailureReason:e,Class:"",InIf:n.inConditional,ElapsedTime:t,Methods:[],PrevInvoc:"",ItemMetadata:Object.create(null)}}(r,o,i);o.typeName&&o.sequenceString&&(c.Count=s.length,c.ModelType=o.sequenceString&&"N"!==o.sequenceString?"Sequence":"Frequency",c.Class="NotInModel"!==r?o.typeName:"",c.Methods=s.map(e=>e.insertText),c.PrevInvoc=o.sequenceString,a.entries.forEach((e,t)=>{const o={Id:c.Id,Index:t,Method:t<s.length?e.insertText:""};c.ItemMetadata[e.name]=o,h(n,`adding item metadata ${e.name} => ${JSON.stringify(o)} to completion result`)})),h(n,`added metadata ${JSON.stringify(c)} to completion result`),a.metadata=c}function k(e,t){e.insertText=e.insertText||e.name,e.sortText=e.sortText||e.name,e.name="★ "+e.name,e.sortText=`*${String.fromCharCode("a".charCodeAt(0)+t)} ${e.sortText}`,h(n,`sort text for entry ${e.name} is ${e.sortText}`)}function v(t){const s=n.languageService.getProgram();if(!s)return{failureReason:"NoProgramFound"};h(n,"program found"),t.throwIfCancellationRequested();const a=s.getTypeChecker(),u=s.getSourceFile(o);if(!u)return{failureReason:"NoSourceFileFound"};h(n,"source file found"),t.throwIfCancellationRequested();const l=function(n,t,o){let r=c.findPrecedingToken(e,o,t);r&&o<=r.end&&(e.isIdentifier(r)||c.isKeyword(e,r.kind))&&(r=c.findPrecedingToken(e,r.getFullStart(),t));if(!r||r.kind!==e.SyntaxKind.DotToken)return;const s=r.parent;if(!e.isPropertyAccessExpression(s))return;const a=i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,n,s.expression,t).typeName;if(!a)return;return{isInConditional:C(s),node:s,typeName:a,sf:t}}(a,u,r);if(t.throwIfCancellationRequested(),!l)return{failureReason:"NoNodeFound"};h(n,`determined type: ${l.typeName}`);const f=function(o){const r=[];try{!function o(r,s,c,u){if(s.getStart()>c.getStart()||u.length>=2)return;t.throwIfCancellationRequested();const f=s.getChildren(l.sf);for(let e=f.length-1;e>=0;e--)o(r,f[e],c,u);if(e.isPropertyAccessExpression(s)&&s!=c){i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,a,s.expression,l.sf).typeName===r&&s.name.text&&(h(n,`pushing prior access: ${s.name.text}`),u.push(s.name.text))}}(o,l.sf,l.node,r)}catch(n){if(!(n instanceof N))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException}h(n,`prior accesses: ${r.join(",")}`);const s=[];r.length>1&&s.push(r[1]+"~"+r[0]);r.length>0&&s.push("N~"+r[0]);return s.push("N"),h(n,`sequences: ${s.join(",")}`),s}(l.typeName);return h(n,`sequences: ${f.join(",")}`),{foundNode:l,sequences:f,failureReason:"None"}}function C(n){let t=n;for(;t.kind!==e.SyntaxKind.SourceFile;){const n=t.parent;if(e.isIfStatement(n)&&n.expression===t||e.isConditionalExpression(n)&&n.condition===t||e.isDoStatement(n)&&n.expression===t||e.isWhileStatement(n)&&n.expression===t||e.isForStatement(n)&&n.condition===t)return!0;t=n}return!1}return a})(n,e,l,t,o,r),u.getCompletionEntryDetails=(n,t,o,i,r,s)=>(function(e,n,t,o,i,r,s,a,c){const u=r.startsWith("★ "),l=u?r.slice("★ ".length):r;u&&h(n,`stripped name for recommendation: ${l}`);const f=n.languageService.getCompletionEntryDetails(o,i,l,s,a,c);f&&u&&(f.name="★ "+f.name,h(n,`starred name for recommendation: ${f.name}`));return f})(0,e,0,n,t,o,i,r,s),u},onConfigurationChanged:function(e){t&&t.log(`received configuration command: ${JSON.stringify(e)}`),o(e)}};function o(e){e.modelPath&&(f=e.modelPath),e.returnLookupKey&&(p=e.returnLookupKey),e.timeout&&(m=e.timeout)}}},function(e,n,t){"use strict";n.__esModule=!0;var o,i,r=t(2);!function(e){e[e.DontSkipAny=0]="DontSkipAny",e[e.SkipAny=1]="SkipAny"}(o=n.GetTypeNameOptions||(n.GetTypeNameOptions={})),function(e){e[e.SkipAny=0]="SkipAny",e[e.Any=1]="Any",e[e.AnyCallExpression=2]="AnyCallExpression",e[e.AnyPropertyAccessExpression=3]="AnyPropertyAccessExpression",e[e.AnyElementAccessExpression=4]="AnyElementAccessExpression",e[e.UnionOrIntersection=5]="UnionOrIntersection",e[e.OtherPrimitive=6]="OtherPrimitive",e[e.OtherNonPrimitive=7]="OtherNonPrimitive",e[e.Literal=8]="Literal",e[e.EmptyIdentifier=9]="EmptyIdentifier"}(i=n.SkipTypeReason||(n.SkipTypeReason={})),n.getTypeNameForPythia=function e(n,t,s,a,c){var u=s.getTypeAtLocation(a),l=u.aliasSymbol||u.getSymbol();if(l){if(0!=(l.flags&(n.SymbolFlags.TypeLiteral|n.SymbolFlags.ObjectLiteral)))return{typeName:s.typeToString(u)};var f=s.getFullyQualifiedName(l);if(f&&'"'===f[0]){var d=function(e){var n=e.indexOf('"',1);if(n>=1)return{partInQuotes:e.substring(1,n),rest:e.substring(n+1)};return{}}(f),p=d.partInQuotes,m=d.rest;if(p){var y=p;if(r.isAbsolute(y)){"index"===r.basename(y)&&(y=r.dirname(y));var g=y.lastIndexOf("/node_modules/");if(g>=0){var h=g+"/node_modules/".length;y=y.substring(h)}else y=r.basename(y);f='"'+y+'"'+(m||"")}}}return{typeName:f}}if(0==(u.getFlags()&n.TypeFlags.NonPrimitive)){if(0!=(u.flags&n.TypeFlags.Any)){var S;if(n.isCallExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"()"}:{skipReason:i.AnyCallExpression};if(n.isPropertyAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"."+a.name.text}:{skipReason:i.AnyPropertyAccessExpression};if(n.isElementAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"[]"}:{skipReason:i.AnyElementAccessExpression};if(t===o.SkipAny)return{skipReason:i.SkipAny};if(n.isIdentifier(a)){var x=a.text;return x?{typeName:x}:{skipReason:i.EmptyIdentifier}}return a.kind===n.SyntaxKind.ThisKeyword?{typeName:"this"}:{skipReason:i.Any}}return 0!=(u.flags&n.TypeFlags.StringLike)?{typeName:"string"}:0!=(u.flags&n.TypeFlags.Boolean)?{typeName:"boolean"}:0!=(u.flags&n.TypeFlags.NumberLike)?{typeName:"number"}:0!=(u.flags&n.TypeFlags.UnionOrIntersection)?{skipReason:i.UnionOrIntersection}:{skipReason:i.OtherPrimitive}}return{skipReason:i.OtherNonPrimitive}}},function(e,n){e.exports=require("path")},function(e,n){e.exports=require("crypto")},function(e,n){e.exports=require("fs")},function(e,n){e.exports=require("zlib")},function(e,n,t){"use strict";function o(e,n,t){return n.kind===e.SyntaxKind.EndOfFileToken?!!n.jsDoc:0!==n.getWidth(t)}function i(e,n){return e.isJsxText(n)&&(n.containsOnlyTriviaWhiteSpaces||n.containsOnlyWhiteSpaces)}function r(e,n,t,r){for(let s=t-1;s>=0;s--){if(i(e,n[s]));else if(o(e,n[s],r))return n[s]}}function s(e,n,t){if(a(e,n))return n;const o=n.getChildren(t),i=r(e,o,o.length,t);return i&&s(e,i,t)}function a(e,n){return e.isToken(n)&&!i(e,n)}Object.defineProperty(n,"__esModule",{value:!0}),n.isInString=function(e,n,t){const o=function(e,n,t){let o=n;e:for(;;){for(const i of o.getChildren(n)){if(i.getFullStart()>t)break;const n=i.getEnd();if(t<n||t===n&&i.kind===e.SyntaxKind.EndOfFileToken){o=i;continue e}}return o}}(e,n,t);if(o&&(e.isStringLiteral(o)||e.isTemplateLiteralToken(o))){const e=o.getStart(n),i=o.getEnd();if(e<t&&t<i)return!0;if(t===i)return!!o.isUnterminated}},n.findPrecedingToken=function(e,n,t,c,u){return function c(l){if(a(e,l)&&l.kind!==e.SyntaxKind.EndOfFileToken)return l;const f=l.getChildren(t);for(let a=0;a<f.length;a++){const l=f[a];if(n<l.end){if(l.getStart(t,!u)>=n||!o(e,l,t)||i(e,l)){const n=r(e,f,a,t);return n&&s(e,n,t)}return c(l)}}const d=r(e,f,f.length,t);return d&&s(e,d,t)}(c||t)},n.isKeyword=function(e,n){return e.SyntaxKind.FirstKeyword<=n&&n<=e.SyntaxKind.LastKeyword}}]);
// SIG // Begin signature block
// SIG // MIIkVQYJKoZIhvcNAQcCoIIkRjCCJEICAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // AG3C77A11vHWhLsNmDkfq25xWIkt5xfJ19BV4+/RMXOg
// SIG // gg2BMIIF/zCCA+egAwIBAgITMwAAAVGejY9AcaMOQQAA
// SIG // AAABUTANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTE5MDUwMjIxMzc0NloX
// SIG // DTIwMDUwMjIxMzc0NlowdDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEeMBwGA1UEAxMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
// SIG // lVrGhmlHHTQe8VXDZnX2YlQYWBRnJ/CjKsYDSLzmVjR/
// SIG // SWEC7oR4ZieUViEstaHst807sai25BwHZm3lPDRTKOPT
// SIG // 7+9TICEvSBvxLasDh+7qWp/pSKujTnMOXzujrPtdkdEN
// SIG // vDMxp/t8uxdpig56KVbtLBLn8uOd4Mc9ejPGwMOPkF7r
// SIG // +/n0fVs0SdgqVOtsECmIhUDH3sOlYeX7j6F5aDd5OvkJ
// SIG // c+84HE+GEZsc8e4zFaT/7ryurGXAcUN1oAf1pMlx4MWm
// SIG // NfSAMy6tkIj4l9mK8okeRLGAat+QT+1NeQ5WbaUrNsCG
// SIG // E5JAwcYTySAyYKMGbuRsoR3aq7Ldzo5EkQIDAQABo4IB
// SIG // fjCCAXowHwYDVR0lBBgwFgYKKwYBBAGCN0wIAQYIKwYB
// SIG // BQUHAwMwHQYDVR0OBBYEFFeCGq5Kue3rGoLuhVAW9u9G
// SIG // Yo3PMFAGA1UdEQRJMEekRTBDMSkwJwYDVQQLEyBNaWNy
// SIG // b3NvZnQgT3BlcmF0aW9ucyBQdWVydG8gUmljbzEWMBQG
// SIG // A1UEBRMNMjMwMDEyKzQ1NDEzNTAfBgNVHSMEGDAWgBRI
// SIG // bmTlUAXTgqoXNzcitW2oynUClTBUBgNVHR8ETTBLMEmg
// SIG // R6BFhkNodHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtp
// SIG // b3BzL2NybC9NaWNDb2RTaWdQQ0EyMDExXzIwMTEtMDct
// SIG // MDguY3JsMGEGCCsGAQUFBwEBBFUwUzBRBggrBgEFBQcw
// SIG // AoZFaHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3BraW9w
// SIG // cy9jZXJ0cy9NaWNDb2RTaWdQQ0EyMDExXzIwMTEtMDct
// SIG // MDguY3J0MAwGA1UdEwEB/wQCMAAwDQYJKoZIhvcNAQEL
// SIG // BQADggIBAFoPgK0uAJ6uyq6ILJSEJ6DB0l1D6/0ajiIS
// SIG // V/t9jm7mNzBb3ZURJW0rEica0cvzggmXUrHvn9gKkPhf
// SIG // 9mmMAQ5lEG3jAfJ9KWC2Zzj4n6nu/EQR9n1WbL18cn4s
// SIG // 2x7m1lqFHzVvxSZWZS18CZhdwaC/BNqdPSt4WoMM/6LX
// SIG // CrUNfkOPg2jmF1pXqiayVLJx7PVIi3K1RdJXi/0NVeW7
// SIG // IaG9jk5WatKs0nazAy1nYcq1DsZ2fqI2e1HU2OFZwrqI
// SIG // G2fWbPbMiW4O2VEvUlYGJKMEbFr+Y1eJW6kw/rRuBc4T
// SIG // sHEUMHW+gPM6djZ3frO1XQrmBbBoCONbnA/KMVX3ADIx
// SIG // fWF+TdrOJpbZKp1Ht/4bVa58SigwMEmJYmrsdi+4CsPm
// SIG // w9ZBGf+aMy0Zyd44w5KJk6z3LaJGPymmdarm9DVJ5jih
// SIG // /t8VCv6ZViSvATkGLlO4CmB+2MvVkijZT+6So+ouNe/t
// SIG // zWv36yJ45wZCkQif3mE7wR/rOYLns6X0FrVOGzaP4/EM
// SIG // Nr6U0PcO35YR+/EZsHd9ffmE5ob+03MQ21pcrXmo+ESt
// SIG // sJN6WNaEs9iTxOTzbjZnfRpn+qHj2YMuyIvSSy6vEp1C
// SIG // 1e2/iD9FF0WPavhnUYzNgBF+prGb7zwiKBtGmB0LvcGb
// SIG // ChCto6r1ovP8XXnnkiPRTeitcOKFUDODMIIHejCCBWKg
// SIG // AwIBAgIKYQ6Q0gAAAAAAAzANBgkqhkiG9w0BAQsFADCB
// SIG // iDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0
// SIG // b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1p
// SIG // Y3Jvc29mdCBDb3Jwb3JhdGlvbjEyMDAGA1UEAxMpTWlj
// SIG // cm9zb2Z0IFJvb3QgQ2VydGlmaWNhdGUgQXV0aG9yaXR5
// SIG // IDIwMTEwHhcNMTEwNzA4MjA1OTA5WhcNMjYwNzA4MjEw
// SIG // OTA5WjB+MQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2Fz
// SIG // aGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UE
// SIG // ChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSgwJgYDVQQD
// SIG // Ex9NaWNyb3NvZnQgQ29kZSBTaWduaW5nIFBDQSAyMDEx
// SIG // MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA
// SIG // q/D6chAcLq3YbqqCEE00uvK2WCGfQhsqa+laUKq4Bjga
// SIG // BEm6f8MMHt03a8YS2AvwOMKZBrDIOdUBFDFC04kNeWSH
// SIG // fpRgJGyvnkmc6Whe0t+bU7IKLMOv2akrrnoJr9eWWcpg
// SIG // GgXpZnboMlImEi/nqwhQz7NEt13YxC4Ddato88tt8zpc
// SIG // oRb0RrrgOGSsbmQ1eKagYw8t00CT+OPeBw3VXHmlSSnn
// SIG // Db6gE3e+lD3v++MrWhAfTVYoonpy4BI6t0le2O3tQ5GD
// SIG // 2Xuye4Yb2T6xjF3oiU+EGvKhL1nkkDstrjNYxbc+/jLT
// SIG // swM9sbKvkjh+0p2ALPVOVpEhNSXDOW5kf1O6nA+tGSOE
// SIG // y/S6A4aN91/w0FK/jJSHvMAhdCVfGCi2zCcoOCWYOUo2
// SIG // z3yxkq4cI6epZuxhH2rhKEmdX4jiJV3TIUs+UsS1Vz8k
// SIG // A/DRelsv1SPjcF0PUUZ3s/gA4bysAoJf28AVs70b1FVL
// SIG // 5zmhD+kjSbwYuER8ReTBw3J64HLnJN+/RpnF78IcV9uD
// SIG // jexNSTCnq47f7Fufr/zdsGbiwZeBe+3W7UvnSSmnEyim
// SIG // p31ngOaKYnhfsi+E11ecXL93KCjx7W3DKI8sj0A3T8Hh
// SIG // hUSJxAlMxdSlQy90lfdu+HggWCwTXWCVmj5PM4TasIgX
// SIG // 3p5O9JawvEagbJjS4NaIjAsCAwEAAaOCAe0wggHpMBAG
// SIG // CSsGAQQBgjcVAQQDAgEAMB0GA1UdDgQWBBRIbmTlUAXT
// SIG // gqoXNzcitW2oynUClTAZBgkrBgEEAYI3FAIEDB4KAFMA
// SIG // dQBiAEMAQTALBgNVHQ8EBAMCAYYwDwYDVR0TAQH/BAUw
// SIG // AwEB/zAfBgNVHSMEGDAWgBRyLToCMZBDuRQFTuHqp8cx
// SIG // 0SOJNDBaBgNVHR8EUzBRME+gTaBLhklodHRwOi8vY3Js
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpL2NybC9wcm9kdWN0cy9N
// SIG // aWNSb29DZXJBdXQyMDExXzIwMTFfMDNfMjIuY3JsMF4G
// SIG // CCsGAQUFBwEBBFIwUDBOBggrBgEFBQcwAoZCaHR0cDov
// SIG // L3d3dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0cy9NaWNS
// SIG // b29DZXJBdXQyMDExXzIwMTFfMDNfMjIuY3J0MIGfBgNV
// SIG // HSAEgZcwgZQwgZEGCSsGAQQBgjcuAzCBgzA/BggrBgEF
// SIG // BQcCARYzaHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3Br
// SIG // aW9wcy9kb2NzL3ByaW1hcnljcHMuaHRtMEAGCCsGAQUF
// SIG // BwICMDQeMiAdAEwAZQBnAGEAbABfAHAAbwBsAGkAYwB5
// SIG // AF8AcwB0AGEAdABlAG0AZQBuAHQALiAdMA0GCSqGSIb3
// SIG // DQEBCwUAA4ICAQBn8oalmOBUeRou09h0ZyKbC5YR4WOS
// SIG // mUKWfdJ5DJDBZV8uLD74w3LRbYP+vj/oCso7v0epo/Np
// SIG // 22O/IjWll11lhJB9i0ZQVdgMknzSGksc8zxCi1LQsP1r
// SIG // 4z4HLimb5j0bpdS1HXeUOeLpZMlEPXh6I/MTfaaQdION
// SIG // 9MsmAkYqwooQu6SpBQyb7Wj6aC6VoCo/KmtYSWMfCWlu
// SIG // WpiW5IP0wI/zRive/DvQvTXvbiWu5a8n7dDd8w6vmSiX
// SIG // mE0OPQvyCInWH8MyGOLwxS3OW560STkKxgrCxq2u5bLZ
// SIG // 2xWIUUVYODJxJxp/sfQn+N4sOiBpmLJZiWhub6e3dMNA
// SIG // BQamASooPoI/E01mC8CzTfXhj38cbxV9Rad25UAqZaPD
// SIG // XVJihsMdYzaXht/a8/jyFqGaJ+HNpZfQ7l1jQeNbB5yH
// SIG // PgZ3BtEGsXUfFL5hYbXw3MYbBL7fQccOKO7eZS/sl/ah
// SIG // XJbYANahRr1Z85elCUtIEJmAH9AAKcWxm6U/RXceNcbS
// SIG // oqKfenoi+kiVH6v7RyOA9Z74v2u3S5fi63V4GuzqN5l5
// SIG // GEv/1rMjaHXmr/r8i+sLgOppO6/8MO0ETI7f33VtY5E9
// SIG // 0Z1WTk+/gFcioXgRMiF670EKsT/7qMykXcGhiJtXcVZO
// SIG // SEXAQsmbdlsKgEhr/Xmfwb1tbWrJUnMTDXpQzTGCFiww
// SIG // ghYoAgEBMIGVMH4xCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xKDAm
// SIG // BgNVBAMTH01pY3Jvc29mdCBDb2RlIFNpZ25pbmcgUENB
// SIG // IDIwMTECEzMAAAFRno2PQHGjDkEAAAAAAVEwDQYJYIZI
// SIG // AWUDBAIBBQCgga4wGQYJKoZIhvcNAQkDMQwGCisGAQQB
// SIG // gjcCAQQwHAYKKwYBBAGCNwIBCzEOMAwGCisGAQQBgjcC
// SIG // ARUwLwYJKoZIhvcNAQkEMSIEIHYAeURLPkPCEp5Rsu7k
// SIG // vr7cBZzUB9ajc+x5uIxcoOpRMEIGCisGAQQBgjcCAQwx
// SIG // NDAyoBSAEgBNAGkAYwByAG8AcwBvAGYAdKEagBhodHRw
// SIG // Oi8vd3d3Lm1pY3Jvc29mdC5jb20wDQYJKoZIhvcNAQEB
// SIG // BQAEggEAfYOYmoBbeT34HzxN8PhRwpQH53MwHVyr6J5/
// SIG // e4n6ozXByuP9Fw8QchBVHlVgRNOR4eirRFcAMlAhuiAr
// SIG // vwEXs1MfBAXU4E+2LkwOiRkin5tMlisLjvBaqQ3GBeeq
// SIG // Ku6Sh22zlu76cN7QUYYoq2qAmxwx8PNh78P8V9CkN89F
// SIG // 6O3mM0SXKW3PYK36HdLPk76XlslAxHLIujtJw5gQlmkB
// SIG // cslOGfntvQa28I3QPgmdzpJkt/WcmP7eOoF1FeHPjwpj
// SIG // 3sR8TAc8Dqo13oAGa6UNEq5PZd19bvCtozIwfGTJr3IO
// SIG // 9hTLBM+GXkV3Q//HtTCmDrTNpvD+ZMU0i5Fze3wWraGC
// SIG // E7YwghOyBgorBgEEAYI3AwMBMYITojCCE54GCSqGSIb3
// SIG // DQEHAqCCE48wghOLAgEDMQ8wDQYJYIZIAWUDBAIBBQAw
// SIG // ggFXBgsqhkiG9w0BCRABBKCCAUYEggFCMIIBPgIBAQYK
// SIG // KwYBBAGEWQoDATAxMA0GCWCGSAFlAwQCAQUABCC2exKh
// SIG // SIbAfD96uLJ4KnQvSfGt8dRtSF+6Nu6yZebKTwIGXYje
// SIG // g+60GBIyMDE5MTAxODIzNDcwOS45OFowBwIBAYACAfSg
// SIG // gdSkgdEwgc4xCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpX
// SIG // YXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYD
// SIG // VQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xKTAnBgNV
// SIG // BAsTIE1pY3Jvc29mdCBPcGVyYXRpb25zIFB1ZXJ0byBS
// SIG // aWNvMSYwJAYDVQQLEx1UaGFsZXMgVFNTIEVTTjpDMEY0
// SIG // LTMwODYtREVGODElMCMGA1UEAxMcTWljcm9zb2Z0IFRp
// SIG // bWUtU3RhbXAgU2VydmljZaCCDx8wggT1MIID3aADAgEC
// SIG // AhMzAAABAYE+Ikb9+jjiAAAAAAEBMA0GCSqGSIb3DQEB
// SIG // CwUAMHwxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNo
// SIG // aW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQK
// SIG // ExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMT
// SIG // HU1pY3Jvc29mdCBUaW1lLVN0YW1wIFBDQSAyMDEwMB4X
// SIG // DTE5MDkwNjIwNDExNVoXDTIwMTIwNDIwNDExNVowgc4x
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xKTAnBgNVBAsTIE1pY3Jv
// SIG // c29mdCBPcGVyYXRpb25zIFB1ZXJ0byBSaWNvMSYwJAYD
// SIG // VQQLEx1UaGFsZXMgVFNTIEVTTjpDMEY0LTMwODYtREVG
// SIG // ODElMCMGA1UEAxMcTWljcm9zb2Z0IFRpbWUtU3RhbXAg
// SIG // U2VydmljZTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
// SIG // AQoCggEBALJQ9ljIhOQ1vaoirU/dKF174ImUd129yLG+
// SIG // eHgBkpHX8sIl1o6pbIG2x5hDXhLqj3X5Z9dLOZkXL3iX
// SIG // rubshmFQsA8AvX7UOw98Nrm88bacNQpMxojid+X87mGk
// SIG // 7UYrQmW5049KH67aU693+5jzkWumJi+CGYd8OS3sZQL7
// SIG // ET2MM5FOXdz7FAvba5Sa7TVY/m55GwZJ7FqAXAnyQAtI
// SIG // hq83x5JKl0CPaBdj4vW6Hex0Gwn6l/iupt8zVW4hk7s9
// SIG // wC3TJCB+/4/ozMDJ9vwwgLgQjIf9Ke27YmQGzCDJMJb5
// SIG // FGhYberjNkCB6zwFTutocnWaO+YvXF0H9ot5jpXEr1kC
// SIG // AwEAAaOCARswggEXMB0GA1UdDgQWBBS/vpia7038DoyY
// SIG // YJEtjBUkHbmkkTAfBgNVHSMEGDAWgBTVYzpcijGQ80N7
// SIG // fEYbxTNoWoVtVTBWBgNVHR8ETzBNMEugSaBHhkVodHRw
// SIG // Oi8vY3JsLm1pY3Jvc29mdC5jb20vcGtpL2NybC9wcm9k
// SIG // dWN0cy9NaWNUaW1TdGFQQ0FfMjAxMC0wNy0wMS5jcmww
// SIG // WgYIKwYBBQUHAQEETjBMMEoGCCsGAQUFBzAChj5odHRw
// SIG // Oi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtpL2NlcnRzL01p
// SIG // Y1RpbVN0YVBDQV8yMDEwLTA3LTAxLmNydDAMBgNVHRMB
// SIG // Af8EAjAAMBMGA1UdJQQMMAoGCCsGAQUFBwMIMA0GCSqG
// SIG // SIb3DQEBCwUAA4IBAQAfPbZa3p/lV7ZYkE/RWfWq6gSj
// SIG // IJ7/jFTgwEpBUefyTHPnQG3BF29rq0jokOqieUqDRiwY
// SIG // Xg6WmnmxwyfmVsSCARdJXfbHWC3wOJw0H0KfULImqZ0D
// SIG // xBYoDg8TXLVVgN/Qgl8gpYzn0n/2QNv7nEVWXlpINueZ
// SIG // 0qhGf7Ghthpy40h99LFC8mmK/O+ceKDRHYZIYeNN8jaQ
// SIG // klRBLWqRLd2mmDSvRxfN4VHcFe/Thx83KzvWvUjaOk3z
// SIG // Gf50OmQpyujU4EKIOy1PKmIbKGrLEvZ/EvpxLbBhsHRS
// SIG // 4tcC37Ph23th4O+wEK/bg55gNHv0U9IjwQjLx8zsu6FF
// SIG // UlTfa31gMIIGcTCCBFmgAwIBAgIKYQmBKgAAAAAAAjAN
// SIG // BgkqhkiG9w0BAQsFADCBiDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEyMDAGA1UEAxMpTWljcm9zb2Z0IFJvb3QgQ2VydGlm
// SIG // aWNhdGUgQXV0aG9yaXR5IDIwMTAwHhcNMTAwNzAxMjEz
// SIG // NjU1WhcNMjUwNzAxMjE0NjU1WjB8MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSYwJAYDVQQDEx1NaWNyb3NvZnQgVGltZS1T
// SIG // dGFtcCBQQ0EgMjAxMDCCASIwDQYJKoZIhvcNAQEBBQAD
// SIG // ggEPADCCAQoCggEBAKkdDbx3EYo6IOz8E5f1+n9plGt0
// SIG // VBDVpQoAgoX77XxoSyxfxcPlYcJ2tz5mK1vwFVMnBDEf
// SIG // QRsalR3OCROOfGEwWbEwRA/xYIiEVEMM1024OAizQt2T
// SIG // rNZzMFcmgqNFDdDq9UeBzb8kYDJYYEbyWEeGMoQedGFn
// SIG // kV+BVLHPk0ySwcSmXdFhE24oxhr5hoC732H8RsEnHSRn
// SIG // EnIaIYqvS2SJUGKxXf13Hz3wV3WsvYpCTUBR0Q+cBj5n
// SIG // f/VmwAOWRH7v0Ev9buWayrGo8noqCjHw2k4GkbaICDXo
// SIG // eByw6ZnNPOcvRLqn9NxkvaQBwSAJk3jN/LzAyURdXhac
// SIG // AQVPIk0CAwEAAaOCAeYwggHiMBAGCSsGAQQBgjcVAQQD
// SIG // AgEAMB0GA1UdDgQWBBTVYzpcijGQ80N7fEYbxTNoWoVt
// SIG // VTAZBgkrBgEEAYI3FAIEDB4KAFMAdQBiAEMAQTALBgNV
// SIG // HQ8EBAMCAYYwDwYDVR0TAQH/BAUwAwEB/zAfBgNVHSME
// SIG // GDAWgBTV9lbLj+iiXGJo0T2UkFvXzpoYxDBWBgNVHR8E
// SIG // TzBNMEugSaBHhkVodHRwOi8vY3JsLm1pY3Jvc29mdC5j
// SIG // b20vcGtpL2NybC9wcm9kdWN0cy9NaWNSb29DZXJBdXRf
// SIG // MjAxMC0wNi0yMy5jcmwwWgYIKwYBBQUHAQEETjBMMEoG
// SIG // CCsGAQUFBzAChj5odHRwOi8vd3d3Lm1pY3Jvc29mdC5j
// SIG // b20vcGtpL2NlcnRzL01pY1Jvb0NlckF1dF8yMDEwLTA2
// SIG // LTIzLmNydDCBoAYDVR0gAQH/BIGVMIGSMIGPBgkrBgEE
// SIG // AYI3LgMwgYEwPQYIKwYBBQUHAgEWMWh0dHA6Ly93d3cu
// SIG // bWljcm9zb2Z0LmNvbS9QS0kvZG9jcy9DUFMvZGVmYXVs
// SIG // dC5odG0wQAYIKwYBBQUHAgIwNB4yIB0ATABlAGcAYQBs
// SIG // AF8AUABvAGwAaQBjAHkAXwBTAHQAYQB0AGUAbQBlAG4A
// SIG // dAAuIB0wDQYJKoZIhvcNAQELBQADggIBAAfmiFEN4sbg
// SIG // mD+BcQM9naOhIW+z66bM9TG+zwXiqf76V20ZMLPCxWbJ
// SIG // at/15/B4vceoniXj+bzta1RXCCtRgkQS+7lTjMz0YBKK
// SIG // dsxAQEGb3FwX/1z5Xhc1mCRWS3TvQhDIr79/xn/yN31a
// SIG // PxzymXlKkVIArzgPF/UveYFl2am1a+THzvbKegBvSzBE
// SIG // JCI8z+0DpZaPWSm8tv0E4XCfMkon/VWvL/625Y4zu2Jf
// SIG // mttXQOnxzplmkIz/amJ/3cVKC5Em4jnsGUpxY517IW3D
// SIG // nKOiPPp/fZZqkHimbdLhnPkd/DjYlPTGpQqWhqS9nhqu
// SIG // BEKDuLWAmyI4ILUl5WTs9/S/fmNZJQ96LjlXdqJxqgaK
// SIG // D4kWumGnEcua2A5HmoDF0M2n0O99g/DhO3EJ3110mCII
// SIG // YdqwUB5vvfHhAN/nMQekkzr3ZUd46PioSKv33nJ+YWtv
// SIG // d6mBy6cJrDm77MbL2IK0cs0d9LiFAR6A+xuJKlQ5slva
// SIG // yA1VmXqHczsI5pgt6o3gMy4SKfXAL1QnIffIrE7aKLix
// SIG // qduWsqdCosnPGUFN4Ib5KpqjEWYw07t0MkvfY3v1mYov
// SIG // G8chr1m1rtxEPJdQcdeh0sVV42neV8HR3jDA/czmTfsN
// SIG // v11P6Z0eGTgvvM9YBS7vDaBQNdrvCScc1bN+NR4Iuto2
// SIG // 29Nfj950iEkSoYIDrTCCApUCAQEwgf6hgdSkgdEwgc4x
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xKTAnBgNVBAsTIE1pY3Jv
// SIG // c29mdCBPcGVyYXRpb25zIFB1ZXJ0byBSaWNvMSYwJAYD
// SIG // VQQLEx1UaGFsZXMgVFNTIEVTTjpDMEY0LTMwODYtREVG
// SIG // ODElMCMGA1UEAxMcTWljcm9zb2Z0IFRpbWUtU3RhbXAg
// SIG // U2VydmljZaIlCgEBMAkGBSsOAwIaBQADFQAIJVNA4iCB
// SIG // +g3btv5VfN3hJ98F8aCB3jCB26SB2DCB1TELMAkGA1UE
// SIG // BhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNV
// SIG // BAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBD
// SIG // b3Jwb3JhdGlvbjEpMCcGA1UECxMgTWljcm9zb2Z0IE9w
// SIG // ZXJhdGlvbnMgUHVlcnRvIFJpY28xJzAlBgNVBAsTHm5D
// SIG // aXBoZXIgTlRTIEVTTjo0REU5LTBDNUUtM0UwOTErMCkG
// SIG // A1UEAxMiTWljcm9zb2Z0IFRpbWUgU291cmNlIE1hc3Rl
// SIG // ciBDbG9jazANBgkqhkiG9w0BAQUFAAIFAOFUlrswIhgP
// SIG // MjAxOTEwMTkwMzUzMzFaGA8yMDE5MTAyMDAzNTMzMVow
// SIG // dDA6BgorBgEEAYRZCgQBMSwwKjAKAgUA4VSWuwIBADAH
// SIG // AgEAAgII6zAHAgEAAgIbTTAKAgUA4VXoOwIBADA2Bgor
// SIG // BgEEAYRZCgQCMSgwJjAMBgorBgEEAYRZCgMBoAowCAIB
// SIG // AAIDFuNgoQowCAIBAAIDB6EgMA0GCSqGSIb3DQEBBQUA
// SIG // A4IBAQAx5FWJZdMgiLbcz7s1hZGi17tjKzIed1KSPtL0
// SIG // 1lQ3iaEU/Tjdd9PcUPy0jVo8NGUD0UUzzX38m5T1GJBs
// SIG // XZQxmBbB4yRLZjspvj5f0zE9d8aYIxg8LXmsq8TxdV0L
// SIG // DF9/UQJ3ujMDQkTIS/27QbkZLQzRYhaQN3oNtq3PZm//
// SIG // LPxKfJLEa64w5YAMVJrXDzcuRo1yI7cO8OuAYK5fop4o
// SIG // bG2frxtvQRlg30z+j/Lu7To8gMg1xhaUysc7YCKcKiEL
// SIG // m526zWtgwj9sLx9BflBuIjtDTdJfPiA8EfpYZgsYfiHS
// SIG // iWYC0tOqS1wvDaWXWie3ny9vsOhkKhcmh+gahzPtMYIC
// SIG // 9TCCAvECAQEwgZMwfDELMAkGA1UEBhMCVVMxEzARBgNV
// SIG // BAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQx
// SIG // HjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEm
// SIG // MCQGA1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENB
// SIG // IDIwMTACEzMAAAEBgT4iRv36OOIAAAAAAQEwDQYJYIZI
// SIG // AWUDBAIBBQCgggEyMBoGCSqGSIb3DQEJAzENBgsqhkiG
// SIG // 9w0BCRABBDAvBgkqhkiG9w0BCQQxIgQgKtQgHwQ5//WA
// SIG // ZyAS+d0Z6qjJ2oAgmrkWhy4sSTGXZEMwgeIGCyqGSIb3
// SIG // DQEJEAIMMYHSMIHPMIHMMIGxBBQIJVNA4iCB+g3btv5V
// SIG // fN3hJ98F8TCBmDCBgKR+MHwxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1w
// SIG // IFBDQSAyMDEwAhMzAAABAYE+Ikb9+jjiAAAAAAEBMBYE
// SIG // FP1beANzy1801XlkGpdeK6S3/OUeMA0GCSqGSIb3DQEB
// SIG // CwUABIIBACp9DtY7t3RpWHXL45bSYh3F/avW/kMZqHiU
// SIG // XxI8UKnLiIZdczIrZuT/Wwiiyb0OUeWTgfnb0yNkTz4P
// SIG // AEqLXCnNOX8xM3L6AqJBu6gPAcUsGOvWVNc4k+BAFpPd
// SIG // BBDzabePq9PxG5ZJh6fpUavdODcM32kfmUt6OmOrcS7S
// SIG // R4Vn3pdozCC2Vd6manXkqBfviUTVrah4tjgDRv9l8YHY
// SIG // 8PgGBeiGYeHOEuQA7qVcb0QASdyGo/Gk2Zzjh8RJSTvB
// SIG // exMtzL5UDL9d3C6vKPJPtGAQB7et9aGa9Xb8hDG6mtn3
// SIG // DIRyqIfwub9m24r7VLJpa1U5k9IqAUUN304s3hal9yQ=
// SIG // End signature block
